src/odemis/driver/orsay.py

```diff
@@ -0,0 +1,766 @@
+                            "datamodel.HybridPlatform.PumpingSystem.Manometer1.Pressure. Parameter passed is %s"
+                            % parameter.Name)
+        elif attributeName == "Actual":
+            self.nitrogenPressure._set_value(float(parameter.Actual), force_write=True)
+
+    def terminate(self):
+        """
+        Called when Odemis is closed
+        """
+        self._system.Manometer1.ErrorState.Unsubscribe(self._updateErrorState)
+        self._system.TurboPump1.ErrorState.Unsubscribe(self._updateErrorState)
+        self._system.TurboPump1.Speed.Unsubscribe(self._updateSpeed)
+        self._system.TurboPump1.Temperature.Unsubscribe(self._updateTemperature)
+        self._system.TurboPump1.Power.Unsubscribe(self._updatePower)
+        self._system.TurboPump1.SpeedReached.Unsubscribe(self._updateSpeedReached)
+        self._system.TurboPump1.IsOn.Unsubscribe(self._updateTurboPumpOn)
+        self.parent.datamodel.HybridPlatform.PrimaryPumpState.Unsubscribe(self._updatePrimaryPumpOn)
+        self._system.Manometer1.Pressure.Unsubscribe(self._updateNitrogenPressure)
+        self._system = None
+
+
+class UPS(model.HwComponent):
+    """
+    This represents the uniterupted power supply from Orsay Physics
+    """
+
+    def __init__(self, name, role, parent, **kwargs):
+        """
+        Defines the following VA's and links them to the callbacks from the Orsay server:
+        • level (FloatVA, read-only, unit is "", value is _system.UPScontroller.BatteryLevel.Actual, between 0 and 1)
+        """
+
+        model.HwComponent.__init__(self, name, role, parent=parent, **kwargs)
+
+        self._system = None
+
+        self.level = model.FloatVA(1.0, readonly=True, unit="")
+
+        self.on_connect()
+
+    def on_connect(self):
+        """
+        Defines the shorthand access points and connects parameter callbacks for the Orsay server.
+        Needs to be called after connection and reconnection to the server.
+        """
+        self._system = self.parent.datamodel.HybridPlatform.UPS
+
+        self._system.UPScontroller.BatteryLevel.Subscribe(self._updateLevel)
+
+        self.update_VAs()
+
+    def update_VAs(self):
+        """
+        Update the VA's. Should be called after reconnection to the server
+        """
+        self._updateLevel()
+
+    def _updateLevel(self, parameter=None, attributeName="Actual"):
+        """
+        Reads the battery level of the UPS from the Orsay server and saves it in the level VA
+        """
+        parameter = self._system.UPScontroller.BatteryLevel if (parameter is None) else parameter
+        if parameter is not self._system.UPScontroller.BatteryLevel:
+            raise Exception("Incorrect parameter passed to _updateLevel. Parameter should be "
+                            "datamodel.HybridPlatform.UPS.UPScontroller.BatteryLevel")
+        elif attributeName == "Actual":
```

Comment:
Note that the `elif` could be replaced by just a `if`, as the previous if always raise an exception. Also, you could even avoid having the main code into a sub-block by turning around the if: 
```
if attributeName != "Actual":
   return
```